kotlin/workflow-core/src/main/java/com/squareup/workflow/Worker.kt

```diff
@@ -0,0 +1,310 @@
+ *
+ * A [Workflow] uses [Worker]s to perform asynchronous work during the render pass by calling
+ * [RenderContext.onWorkerOutput] or [RenderContext.runningWorker]. When equivalent [Worker]s are
+ * passed in subsequent render passes, [doesSameWorkAs] is used to calculate which [Worker]s are
+ * new and should be started, and which ones are continuations from the last render pass and
+ * should be allowed to continue working. [Worker]s that are not included in a render pass are
+ * cancelled.
+ *
+ * ## Example: Network request
+ *
+ * Let's say you have a network service with an API that returns a [Deferred], and you want to
+ * call that service from a [Workflow].
+ *
+ * ```
+ * interface TimeService {
+ *   fun getTime(timezone: String): Deferred<Long>
+ * }
+ * ```
+ *
+ * The first step is to define a [Worker] that can call this service, and maybe an extension
+ * function on your service class:
+ * ```
+ * fun TimeService.getTimeWorker(timezone: String): Worker<Long> = TimeWorker(timezone, this)
+ *
+ * private class TimeWorker(
+ *   val timezone: String,
+ *   val service: TimeService
+ * ): Worker<Long> {
+ *
+ *   override suspend fun performWork(emitter: Emitter<Long>) {
+ *     val timeDeferred = service.getTime(timezone)
+ *     emitter.emitOutput(timeDeferred.await())
+ *   }
+ * }
+ * ```
+ *
+ * You also need to define how to determine if a previous [Worker] is already doing the same work.
+ * This will ensure that if the same request is made by the same [Workflow] in adjacent render
+ * passes, we'll keep the request alive from the first pass.
+ * ```
+ *   override fun doesSameWorkAs(otherWorker: Worker<*>): Boolean =
+ *     otherWorker is TimeWorker &&
+ *       timezone == otherWorker.timezone
+ * ```
+ *
+ * Now you can request the time from your [Workflow]:
+ * ```
+ * class MyWorkflow(private val timeWorker: TimeWorker) {
+ *   override fun render(…): Foo {
+ *     context.onWorkerOutput(timeWorker) { time -> emitOutput("The time is $time") }
+ *   }
+ * ```
+ *
+ * Alternatively, if the response is a unique type, unlikely to be shared by any other workers,
+ * you don't even need to create your own [Worker] class, you can use a builder, and the worker
+ * will automatically be distinguished by that response type:
+ * ```
+ * interface TimeService {
+ *   fun getTime(timezone: String): Deferred<TimeResponse>
+ * }
+ *
+ * fun TimeService.getTimeWorker(timezone: String): Worker<TimeResponse> =
+ *   Worker.from { getTime(timezone).await()) }
+ * ```
+ *
+ * @see create
+ * @see from
+ * @see fromNullable
+ * @see fromChannel
+ * @see Deferred.asWorker
+ * @see BroadcastChannel.asWorker
+ */
+interface Worker<out T> {
+
+  /**
+   * Passed into [performWork] to allow the [Worker] to emit outputs.
+   */
+  interface Emitter<in T> {
+    /**
+     * Delivers an output value to the [Workflow] that is running this [Worker].
+     */
+    suspend fun emitOutput(output: T)
```

Comment:
Why does this need to be a suspend function?

Reply:
Because this usually maps to a `SendChannel.send`, which is a suspending operation. This is how workers handle backpressure.